ts/explorer-apis.ts

```diff
@@ -67,16 +69,118 @@ const getTokenInfosFromBlockscout = async (
+
+const ABI_CODER = new ethers.utils.AbiCoder();
+
+const getDecimals = async (
+  provider: ethers.providers.JsonRpcProvider,
+  tokens: string[]
+): Promise<number[]> => {
+  const Multicall3 = getMulticall3(provider);
+  const decimalsCalldata = new ethers.utils.Interface(
+    erc20Abi
+  ).encodeFunctionData('decimals');
+  const cds = tokens.map((token) => ({
+    target: token,
+    callData: decimalsCalldata,
+  }));
+  const decimalsRet = await Multicall3.tryAggregate(false, cds);
+  const decimals = decimalsRet.map((r: any, idx: number) => {
+    if (!r.success) return 0;
+
+    try {
+      return ABI_CODER.decode(['uint8'], r.returnData)[0];
+    } catch (err) {
+      console.log('errror decoding', r.returnData, tokens[idx]);
+    }
+  }) as number[];
+  return decimals;
+};
+
+export const getTokenInfosFromChain = async (
+  config: IConfig
+): Promise<ITokenInfo[]> => {
+  const provider = new ethers.providers.JsonRpcProvider(config.rpcUrl);
+  const settlement = new ethers.Contract(
+    config.gpv2Settlement,
+    settlementAbi,
+    provider
+  );
+  const tradeFilter = settlement.filters.Trade();
+  const currentBlock = await provider.getBlockNumber();
+  const logs = await settlement.queryFilter(
+    tradeFilter,
+    currentBlock - config.lookbackRange,
+    currentBlock
+  );
+  const allTokens = Array.from(
+    new Set(
+      logs.map((log) => [log.args!.sellToken, log.args!.buyToken]).flat()
+    ).values()
+  ).filter(
+    (x) =>
+      x.toLowerCase() !==
+      '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE'.toLowerCase()
+  );
+  const orderbookApi = getOrderbookApi(config);
+  const naitvePrices = await Promise.allSettled(
+    allTokens.map((token) => orderbookApi.getNativePrice(token))
+  );
+
+  const multiplier = config.network === 'gnosis' ? 1 : await getEthPrice();
+  const allDecimals = await getDecimals(provider, allTokens);
+
+  const allTokenInfo: ITokenInfo[] = allTokens.map((token, idx) => {
+    const nativePriceResponse = naitvePrices[idx];
+    let price: number = 0;
+    const decimals = allDecimals[idx];
+    const denominator = 10 ** (18 - decimals);
+
+    if (nativePriceResponse.status === 'fulfilled') {
+      price =
+        ((nativePriceResponse.value.price || 0) * multiplier) / denominator;
+    } else {
+      // console.log('failure', token, nativePriceResponse.reason);
+    }
+
+    return {
+      address: token,
+      rate: price,
+      decimals,
+      symbol: '',
+    };
+  });
+  return allTokenInfo;
+};
+
 export const getTokenBalances = async (
   address: string,
-  network: IConfig['network']
+  network: IConfig['network'],
+  strategy: 'explorer' | 'chain',
+  config: IConfig
 ): Promise<ITokenInfo[]> => {
-  switch (network) {
-    case 'mainnet': {
-      return getTokenInfosFromEthPlorer(address);
+  switch (strategy) {
+    case 'explorer': {
+      return network === 'mainnet'
+        ? getTokenInfosFromEthPlorer(address)
+        : getTokenInfosFromBlockscout(address);
```

Comment:
nitt: make this an exhaustive switch statement so that when we add a new network we are prompted to think about what explorer to use (instead of defaulting to blockscout which may not exist)